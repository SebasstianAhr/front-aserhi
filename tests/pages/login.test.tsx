import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Login from "../../src/pages/login/login";
import '@testing-library/jest-dom';
import { useNavigate } from "react-router-dom";

// Mock data
jest.mock("../../src/core/mocks/mock-data", () => ({
  users: [
    { identification: "12345", password: "password123" },
    { identification: "67890", password: "password456" },
  ],
}));

// Mock useNavigate
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: jest.fn(),
}));

const renderWithRouter = (component: JSX.Element) => {
  return render(<BrowserRouter>{component}</BrowserRouter>);
};

describe("Login Component", () => {
  test("Renders the initial component correctly", () => {
    renderWithRouter(<Login />);
    expect(screen.getByText(/ingresar/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/identificación/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/contraseña/i)).toBeInTheDocument();
  });

  test("Shows validation errors when fields are empty", async () => {
    renderWithRouter(<Login />);
    fireEvent.click(screen.getByRole("button", { name: /ingresar/i }));

    const errorMessages = screen.getAllByClassName("login__error-message");
    expect(errorMessages[0]).toHaveTextContent(/identificación es obligatoria/i);
    expect(errorMessages[1]).toHaveTextContent(/contraseña es obligatoria/i);
  });

  test("Shows an error when incorrect credentials are entered", async () => {
    renderWithRouter(<Login />);
    fireEvent.change(screen.getByPlaceholderText(/identificación/i), {
      target: { value: "11111" },
    });
    fireEvent.change(screen.getByPlaceholderText(/contraseña/i), {
      target: { value: "wrongpassword" },
    });
    fireEvent.click(screen.getByRole("button", { name: /ingresar/i }));

    expect(await screen.findByText(/identificación o contraseña incorrecta/i)).toBeInTheDocument();
  });

  test("Logs in successfully with valid credentials", async () => {
    const mockNavigate = jest.fn();
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);

    renderWithRouter(<Login />);
    fireEvent.change(screen.getByPlaceholderText(/identificación/i), {
      target: { value: "12345" },
    });
    fireEvent.change(screen.getByPlaceholderText(/contraseña/i), {
      target: { value: "password123" },
    });
    fireEvent.click(screen.getByRole("button", { name: /ingresar/i }));

    // Verificar que no aparece el mensaje de error
    expect(screen.queryByText(/identificación o contraseña incorrecta/i)).not.toBeInTheDocument();

    // Verificar que la redirección ocurre
    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith("/home"); // Ajusta a tu ruta esperada
    });
  });

  test("The '¿Olvidaste tu contraseña?' link redirects correctly", () => {
    renderWithRouter(<Login />);
    const link = screen.getByText(/¿olvidaste tu contraseña\?/i);
    expect(link).toHaveAttribute("href", "/forgotPassword"); // Ajusta la ruta
  });
});
