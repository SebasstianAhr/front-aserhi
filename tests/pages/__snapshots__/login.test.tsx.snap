// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component El componente debe coincidir con el snapshot 1`] = `
<DocumentFragment>
  <div
    class="login__main"
  >
    <div
      class="login__container"
    >
      <div>
        <img
          alt="Logo"
          src="test-file-stub"
        />
      </div>
      <h3>
        INICIAR 
        <span>
          SESIÓN
        </span>
      </h3>
      <div
        class="login__form-container"
      >
        <form
          class="login__form"
        >
          <input
            class="login__input "
            name="identification"
            placeholder="Identificación"
            type="text"
          />
          <input
            class="login__input login__input--bottom "
            name="password"
            placeholder="Contraseña"
            type="password"
          />
          <div
            class="login__form-button"
          >
            <button
              type="submit"
            >
              INGRESAR
            </button>
          </div>
        </form>
        <a
          class="login__link"
          href="/forgot-password"
        >
          <span>
            ¿Olvidaste tu contraseña?
          </span>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
